### PR TITLE
[InstCombine] Remove dead phi web

### PR DESCRIPTION
**Pull request by Chengjunp**
In current visitPHINode function during InstCombine, it can remove dead phi cycles (all phis have one use, which is another phi). However, it cannot deal with the case when the phis form a web (all phis have one or more uses, and all the uses are phi). This change extends the algorithm so that it can also deal with the dead phi web.

### CODE REVIEW

**Review by nikic — APPROVED** [review]
LGTM. This seems like a reasonable extension, and the number of visited phi nodes is still limited to a small number. Compile-time impact looks ok: https://llvm-compile-time-tracker.com/compare.php?from=223e2efa5e886502a9467b7ef700ebce9b7886e8&to=55a38338a2ebeef31ad8da34d965763f522e5799&stat=instructions:u

File: llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp [hunk 1]
```diff
@@ -976,24 +976,26 @@ Instruction *InstCombinerImpl::foldPHIArgOpIntoPHI(PHINode &PN) {
...
+    if (Visited.size() == 16)
+      return false;
+    for (User *Use : Phi->users()) {
+      if (PHINode *PhiUse = dyn_cast<PHINode>(Use))
+        Stack.push_back(PhiUse);
+      else
+        return false;
+    }
+  }
+  return true;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp:998]: Why not just delete the whole dead phi web here? Thus we can avoid `O(n^2)` traversals.

File: llvm/lib/Transforms/InstCombine/InstCombineInternal.h [hunk 2]
```diff
@@ -634,6 +634,11 @@ class LLVM_LIBRARY_VISIBILITY InstCombinerImpl final
   Instruction *foldPHIArgZextsIntoPHI(PHINode &PN);
   Instruction *foldPHIArgIntToPtrToPHI(PHINode &PN);
 
+  // If the phi is within a phi web, which is formed by the def-use chain
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineInternal.h:637]:
```suggestion
  /// If the phi is within a phi web, which is formed by the def-use chain
```
Please use `///` for header comments: https://llvm.org/docs/CodingStandards.html#comment-formatting.

File: llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp [hunk 3]
```diff
@@ -53,6 +53,34 @@ void InstCombinerImpl::PHIArgMergedDebugLoc(Instruction *Inst, PHINode &PN) {
   }
 }
 
+// If the phi is within a phi web, which is formed by the def-use chain
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombinePHI.cpp:56]:
```suggestion
/// If the phi is within a phi web, which is formed by the def-use chain
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**